### PR TITLE
Create CoinFabrik_On_Ink_Integration_Tests.md

### PR DESCRIPTION
**Pull request by valeriacaracciolo**
#### Project Abstract

CoinFabrik On Ink Integration Tests:

During the development of our open source ink! static analyzer tool [Scout](https://github.com/CoinFabrik/scout), we came across several inconsistencies and missing functionalities in ink! integration tests.

Our objective is to conduct a comprehensive analysis to identify any other missing functionalities in integration tests, and to propose and develop new testing features based on our findings. 

Our intention is to `flatten the anvil` of ink! integration testing. With a properly flattened anvil, quality tools can be built.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by valeriacaracciolo** [comment]
Thank you, David. We greatly appreciate your input. We are confident that dedicating a 4-week development cycle will adequately address the less complex scenarios, while a 6-weeks projection will allow us to secure an adequate team. This delineates our proposed approach for optimizing the workflow across these two milestones.

File: applications/CoinFabrik_On_Ink_Integration_Tests.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+### Overview
+
+- **Total Estimated Duration:** 6 weeks
+- **Full-Time Equivalent (FTE):**  2.5 FTE
+(0.25 Project Manager,
+0.25 Tech Lead,
+1 Full time Sr. Rust Developer,
+1 Full Time SemiSr. Rust Developer)
+- **Total Costs:**  U$D 40,500
```
Comment (keeganquigley) [inline applications/CoinFabrik_On_Ink_Integration_Tests.md:112]:
```suggestion
- **Total Costs:**  40,500 USD
```

File: applications/CoinFabrik_On_Ink_Integration_Tests.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+
+### Milestone 1: Analysis
+- **Estimated duration:** 2 weeks
+- **FTE:**  2.75
+- **Costs:** U$D 13,500
```
Comment (keeganquigley) [inline applications/CoinFabrik_On_Ink_Integration_Tests.md:118]:
```suggestion
- **Costs:** 13,500 USD
```

File: applications/CoinFabrik_On_Ink_Integration_Tests.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ **1** | Analyze | Study and compare Integration and E2E (End-to-End) tests in ink!.
+ **2** | Evaluate | Assign a complexity level to each finding based on the difficulty of implementing the missing or enhanced functionality.
+ **3** | Estimate | Indicate which tests shall be developed during the next milestone delivery. If the allocated time for the next milestone is insufficient, we may consider requesting an extension for this grant.
+
+
+
+### Milestone 2: Execution
+- **Estimated duration:** 4 weeks
+- **FTE:** 2.75
+- **Costs:** U$D27,000
```
Comment (keeganquigley) [inline applications/CoinFabrik_On_Ink_Integration_Tests.md:136]:
```suggestion
- **Costs:** 27,000 USD
```

**Review by keeganquigley — CHANGES_REQUESTED** [review]
Thanks for the application @valeriacaracciolo I do tend to agree that it might be better to break up the grant since the timeline isn't known for the execution stage. That way an extension won't be needed. Either that or just extend it now to be on the safe side. I also think that $40k seems a bit expensive for writing only the tests.

- How do you plan to maintain Scout once the grant is completed? Will you go after [alternative sources of funding](https://github.com/w3f/Grants-Program#rocket-alternative-funding-sources)?
- How do you plan to help make your tool become a standard for dev users?
- Can you briefly explain more about the problems you faced in regard to the functions?

**Comment by valeriacaracciolo** [comment]
Thank you @keeganquigley, here we answer your questions. Regards!

## Question 1: How do you plan to maintain Scout once the grant is completed? Will you go after [alternative sources of funding](https://github.com/w3f/Grants-Program#rocket-alternative-funding-sources)?

At this point we are searching for user’s feedback before applying for more funding. But we are certainly considering the DAO Treasury, as suggested by David.

## Question 2: How do you plan to help make your tool become a standard for dev users?

We plan both to grow and maintain scout audit. For example, we have looked for funding from Aleph Zero, who are also users of ink!. There, our intent was twofold: funding the integral growth of the tool, but also and at the same time having increased contact with ink! developers and live projects. Generally speaking, we are putting our effort to have the tool known and used by ink! devs. We’d appreciate any help in getting the word out. We have already been answering issues opened by users, and will continue to do so--supporting the community as it grows. Hopefully we will not be alone in adding detectors, and improving scout audit. To that purpose, we will continue to work on documentation for contributions.

## Question 3: Can you briefly explain more about the problems you faced in regard to the functions?

In the following table, we explain the issues found so far. Generally, a functionality that is available in E2E tests is lacking in integration tests. We provide links to the respective implementations or their absence in each case. Our goal here is to identify other issues like these and consider implementation solutions to bring missing functionalities and improvements to integration tests.
| Issue Number | Issue Name | Description | Feasibility | References |
|--------------|------------|-------------|-------------|------------|
| 1 | Alice and Bob's addresses | Alice and Bob's addresses should match across integration and E2E tests. | Feasible | Accounts on Integration Tests: [Link](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/env/src/engine/off_chain/test_api.rs#L333) <br> Accounts on E2E tests: Addresses are used from this library [Link](https://github.com/paritytech/substrate/blob/28e906dffcaa91e85f59aff628d953ebeb036ae2/primitives/keyring/src/sr25519.rs#L108C38-L108C38) <br> Account addresses are different for both Alice and Bob. |
| 2 | Storage Limitations | The storage in the integration environment should have the same limitations as in the blockchain environment (end-to-end). | Feasible | Set_contract_storage is implemented differently in integration tests vs E2E tests. <br> In integration tests, set_contract_storage is implemented this way: [Link](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/env/src/engine/off_chain/impls.rs#L185) <br> In E2E tests, the implementation calls the ext::set_storage function, this function is implemented in a way that checks maximum value lengths: [Link1](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/env/src/engine/on_chain/impls.rs#L215) [Link2](https://github.com/paritytech/substrate/blob/28e906dffcaa91e85f59aff628d953ebeb036ae2/frame/contracts/src/wasm/runtime.rs#L790) <br> The checking of maximum value lengths through the ext::set_storage function is not implemented in integration tests. |
| 3 | Delegate call | Ability to use delegate calls in integration tests. | Complex, yet feasible. | Not implemented in integration tests: [Link](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/env/src/engine/off_chain/impls.rs#L449) <br> Notice in line 460: “off-chain environment does not support delegated contract invocation" <br> In E2E tests, it is implemented here: [Link](https://github.com/paritytech/substrate/blob/28e906dffcaa91e85f59aff628d953ebeb036ae2/frame/contracts/src/wasm/runtime.rs#L1467) |
| 4 | Cross contract calls | Ability to perform contract-to-contract calls during integration tests. | To be evaluated. | Not implemented in integration tests: [Link](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/env/src/engine/off_chain/impls.rs#L432) <br> Notice in line 446: “off-chain environment does not support contract invocation" <br> In E2E tests, it is implemented here: [Link](https://github.com/paritytech/substrate/blob/28e906dffcaa91e85f59aff628d953ebeb036ae2/frame/contracts/src/wasm/runtime.rs#L1412) |
| 5 | Gas usage | Gas usage in the integration environment should have the same limitations as in the blockchain environment (end-to-end). | To be evaluated. | Not implemented in integration tests: [Link](https://github.com/paritytech/ink/blob/c2af39883aab48c71dc09dac5d06583f2e84dc54/crates/engine/src/ext.rs#L404) <br> Notice in line 405: “off-chain environment does not yet support `gas_left" <br> In E2E tests, it is implemented here: [Link](https://github.com/paritytech/substrate/blob/28e906dffcaa91e85f59aff628d953ebeb036ae2/frame/contracts/src/wasm/runtime.rs#L1974) |

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the detailed reply here. Could you integrate the above already into the specification of the milestone tables? Apart from this, I recommend splitting up the grant or adding more concrete deliverables to milestone 2. At the moment, the outcome of milestone one could theoretically be "nothing found" and milestone 2 "nothing to develop," and technically, you still would have completed the milestone successfully.

**Review by keeganquigley — APPROVED** [review]
Thanks for breaking it up @valeriacaracciolo looks good, and based on your previous work, I'm happy to go forward with it at this stage.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update here. Could you also update the level of the grant at the beginning? And to mention it, if you reduce it to 10k (level 1), it only needs two approvals and would be basically already approved (Keegan and myself).

**Comment by valeriacaracciolo** [comment]
Hello David,

Sorry about that, we have updated the Level. Yet we will need to remain in Level 2, my team won’t allow a budget adjustment at this point. Let’s hope we can move forward.

Best,

**Review by keeganquigley — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I will share it with the rest of the team.

**Review by semuelle — APPROVED** [review]
Happy to proceed.

**Pull request merged**